apps/preview/next/pages/showcases/Combobox/ComboboxBasic.tsx

```diff
@@ -0,0 +1,288 @@
+    value: 'afghanistan',
+  },
+  {
+    label: 'Bahamas',
+    value: 'bahamas',
+  },
+  {
+    label: 'Canada',
+    value: 'canada',
+  },
+  {
+    label: 'Denmark',
+    value: 'denmark',
+  },
+  {
+    label: 'Egypt',
+    value: 'egypt',
+  },
+];
+
+// eslint-disable-next-line no-promise-executor-return
+const delay = () => new Promise((resolve) => setTimeout(resolve, Math.random() * 1000));
+
+const mockAutocompleteRequest = async (phrase: string) => {
+  await delay();
+  const results = options.filter((option) => option.value.toLowerCase().startsWith(phrase.toLowerCase()));
+  return results;
+};
+
+export default function ComboboxBasic() {
+  const inputRef = useRef<HTMLInputElement>(null);
+  const dropdownRef = useRef<HTMLUListElement>(null);
+  const [searchValue, setSearchValue] = useState<string>('');
+  const [selectedValue, setSelectedValue] = useState<string>('');
+  const [isValid, setIsValid] = useState<boolean | undefined>(undefined);
+  const [isLoadingSnippets, setIsLoadingSnippets] = useState(false);
+  const { isOpen, close, open, toggle } = useDisclosure();
+  const [snippets, setSnippets] = useState<{ label: string; value: string }[]>([]);
+  const { refs, style } = useDropdown({ isOpen, onClose: close, placement: 'bottom-start', middleware: [offset(4)] });
+  const [isDisabled, setIsDisabled] = useState(false);
+  const id = useId();
+  const listId = useId();
+
+  const { current: currentFocus, focusables: focusableElements } = useTrapFocus(dropdownRef, {
+    trapTabs: false,
+    arrowKeysOn: true,
+    activeState: isOpen,
+    initialFocus: false,
+  });
+
+  const handleFocusInput = () => {
+    inputRef.current?.focus();
+  };
+
+  const handleReset = () => {
+    setSearchValue('');
+    setSnippets([]);
+    close();
+  };
+
+  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
+    const phrase = event.target.value;
+    setSelectedValue('');
+    if (phrase) {
+      setSearchValue(phrase);
+    } else {
+      handleReset();
+    }
+  };
+
+  const handleBlur = () => {
+    if (isOpen) return;
+    setIsValid(!!selectedValue);
+  };
+
+  const handleInputKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
+    if (event.key === 'Escape') handleReset();
+    if (event.key === 'Enter') close();
+    if (event.key === 'ArrowUp') {
+      open();
+      if (isOpen) {
+        const focusableElementsAmount = focusableElements.length;
+        focusableElements[focusableElementsAmount - 1].focus();
+      }
+    }
+    if (event.key === 'ArrowDown') {
+      open();
+      if (isOpen) {
+        focusableElements[0].focus();
+      }
+    }
+  };
+
+  const selectOption = (event: FormEvent, option: SelectOption) => {
+    setSearchValue(option.label);
+    setSelectedValue(option.label);
+    close();
+    handleFocusInput();
+  };
+
+  const handleOptionItemKeyDown = (event: KeyboardEvent<HTMLButtonElement>, option: SelectOption) => {
+    if (event.key === 'Escape') {
```

Comment:
idea: maybe its worth to define this statement in additional function variable since its used in few places, same goes for other event.key check

Reply:
Do you mean a utility function to handle all event keys? For me, it would be almost the same amount of code, but probably I'm missing something.

Reply:
probably not much gain indeed, then lets leave it as it :)